### PR TITLE
Ephemeral nodes should not get any stateful services

### LINKED ISSUE #2487: stateful: true service should not get deployed on instances with ephemeral[=yes] label

### PR DESCRIPTION
**Pull request by jnummelin**
fixes #2487

### CODE REVIEW

**Comment by jnummelin** [comment]
eh, there's _slight_  problem if there's a service which already has instances scheduled for `ephemeral` nodes:
```
$ k service deploy redis
 [done] Deploying service redis      
⊛ Deployed instance e2e/null/redis-1 to node moby

$ k node label add moby ephemeral=true

$ k service deploy redis
 [fail] Deploying service redis      
 [error] 500 : Cannot find applicable node for service instance e2e/null/redis-1: Did not find any non-ephemeral nodes for stateful service:
```

**Comment by SpComb** [comment]
Hmm right, I was only testing with `strategy: daemon` and then making sure that new services do not get deployed to ephemeral nodes. 

Hmm... odd error though, how many nodes do you have in your grid? If it's just one node, then that error looks sensible? Otherwise for the case of a stateful service instance deployed to a node that gets labled as ephemeral, I would expect it to behave like #2277 for an offline node: `Strategy Scheduler::Strategy::HighAvailability did not find any node: `

Labeling a node as ephemeral after the fact is a bit of an odd use-case, that label should come from the Docker labels when the agent first connects.

**Comment by SpComb** [comment]
> Otherwise for the case of a stateful service instance deployed to a node that gets labled as ephemeral, I would expect it to behave like #2277 for an offline node: Strategy Scheduler::Strategy::HighAvailability did not find any node:

And the workaround should be the same: `kontena service rm --instance` the affected stateful instance on the ephemeral node.

**Comment by jnummelin** [comment]
> Hmm... odd error though, how many nodes do you have in your grid? If it's just one node, then that error looks sensible?

Kinda yes, but still some part of my brain would expect that an existing instance could be re-deployed to the same node where the data already is.

> Labeling a node as ephemeral after the fact is a bit of an odd use-case, that label should come from the Docker labels when the agent first connects.

I agree that this is bit of odd case, but I think there's a real possibility of this happening since < 1.4 releases have not had any deployment controls over ephemeral nodes. The worst part is that the only workaround involves manual removal of instance(s) which could very well lead to data loss.

**Comment by SpComb** [comment]
Indeed:

#### `kontena service show stateful-1/foo`
Stateful service is deployed to `core-02`:

```
development/stateful-1/foo:
  created: 2017-07-03T14:09:12.247Z
  updated: 2017-07-03T14:09:12.247Z
  stack: development/stateful-1
  desired_state: running
  image: redis:latest
  revision: 1
  stack_revision: 1
  stateful: yes
  scaling: 1
  strategy: ha
  stop_grace_period: 10s
  deploy_opts:
  dns: foo.stateful-1.development.kontena.local
  instances:
    stateful-1/foo/1:
      scheduled_to: core-02
      deploy_rev: 2017-07-03 14:09:12 UTC
      rev: 2017-07-03 14:09:12 UTC
      state: running
      containers:
        stateful-1.foo-1 (on core-02):
          dns: foo-1.stateful-1.development.kontena.local
          ip: 10.81.128.108
          public ip: 91.150.10.190
          status: running
```

#### `kontena node label add core-02 ephemeral=true`

Label core-02 as ephemeral

#### `kontena stack deploy stateful-1`

Scheduling fails per #2277 
```
 [done] Triggering deployment of stack stateful-1     
 [fail] Waiting for deployment to start     
        Cannot find applicable node for service instance development/stateful-1/foo-1: Strategy Scheduler::Strategy::HighAvailability did not find any node
 [error] RuntimeError : deploy failed
         See /home/kontena/.kontena/kontena.log or run the command again with environment DEBUG=true set to see the full exception
```

#### `kontena service rm --instance=1 stateful-1/foo`
Force stateful service instance to get re-scheduled:

```
Destructive command. To proceed, type "stateful-1/foo/1" or re-run this command with --force option.
> Enter 'stateful-1/foo/1' to confirm:  stateful-1/foo/1
 [done] Removing service instance stateful-1/foo/1      
```

#### `kontena stack deploy stateful-1`

Able to schedule instance to new node:

```
 [done] Triggering deployment of stack stateful-1     
 [done] Waiting for deployment to start     
 [done] Deploying service foo     
```

#### `kontena service show stateful-1/foo`
```
development/stateful-1/foo:
  created: 2017-07-03T14:09:12.247Z
  updated: 2017-07-03T14:09:12.247Z
  stack: development/stateful-1
  desired_state: running
  image: redis:latest
  revision: 1
  stack_revision: 1
  stateful: yes
  scaling: 1
  strategy: ha
  stop_grace_period: 10s
  deploy_opts:
  dns: foo.stateful-1.development.kontena.local
  instances:
    stateful-1/foo/1:
      scheduled_to: core-01
      deploy_rev: 2017-07-03 14:09:53 UTC
      rev: 2017-07-03 14:09:53 UTC
      state: running
      containers:
        stateful-1.foo-1 (on core-02):
          dns: foo-1.stateful-1.development.kontena.local
          ip: 10.81.128.108
          public ip: 91.150.10.190
          status: running
        stateful-1.foo-1 (on core-01):
          dns: foo-1.stateful-1.development.kontena.local
          ip: 10.81.128.30
          public ip: 91.150.10.190
          status: running
```

The dup container disappers once the agent notices that it should be terminated.

**Comment by jnummelin** [comment]
I _think_  the `Ephemeral` filter should consider an `ephemeral` node being valid deployment target for stateful service IF there's already existing instance on that node.

**Comment by SpComb** [comment]
> I think the Ephemeral filter should consider an ephemeral node being valid deployment target for stateful service IF there's already existing instance on that node.

Then that sounds vaugely like the offline node grace period stickyness, and should be implemented at the strategy level.... although that offline grace period thing is implemented in the `GridScheduler` and does not apply to the actual deploy, which is the exactly the missing thing that makes the resulting error in #2277 so nasty.

**Comment by jnummelin** [comment]
> Then that sounds vaugely like the offline node grace period stickyness, and should be implemented at the strategy level....

The problem is that filters are run first and thus ephemeral nodes will not get to strategy level at all for stateless services. I don't see a problem if the filter would return ephemeral node if it already has an stateful instance, the strategy should then pick that up.

**Comment by jnummelin** [comment]
With the latest commit that allows ephemeral node for scheduling only if it already has an instance:
```
k service create --stateful redis redis
 [done] Creating redis service      

$ k service deploy redis
 [done] Deploying service redis      
⊛ Deployed instance e2e/null/redis-1 to node moby

$ k node label add moby ephemeral=true

$ k service deploy redis
 [done] Deploying service redis      
⊛ Deployed instance e2e/null/redis-1 to node moby

$ k service create --stateful another redis
 [done] Creating another service      

$ k service deploy another
 [fail] Deploying service another      
 [error] 500 : Cannot find applicable node for service instance e2e/null/another-1: Did not find any non-ephemeral nodes for stateful service:
```

**Comment by SpComb** [comment]
> With the latest commit that allows ephemeral node for scheduling only if it already has an instance:

BTW I think offline nodes should also behave similarly re #2277: existing stateful instances should just keep being scheduled to offline nodes, then the deploy will fail on them being offline with an informative error message.

File: docs/using-kontena/nodes.md [hunk 1]
```diff
@@ -74,3 +74,5 @@ Nodes labeled as `ephemeral=yes` will automatically be removed by the Kontena Ma
 The nodes should not be [initial nodes](grids.md#initial-nodes), and they should not have any stateful services deployed on them.
 
 Ephemeral nodes are intended to be used for autoscaled nodes, which may be provisioned automatically, and then cleaned up once terminated.
+
+Ephemeral nodes will not be available for sheduling of stateful services as the service state is only persisted on local disk.
```
Comment (jakolehm) [inline docs/using-kontena/nodes.md]: `sheduling` -> `scheduling`

**Pull request merged**